Make the actor transport an associated type of the DistributedActor protocol

DougGregor opened the pull request:
Eliminate the required use of existentials in distributed actors by
introducing the `Transport` associated type into the
`DistributedActor` protocol. Each distributed actor has a known
(concrete) actor transport type, reducing storage requirements and
transport dynamism when it isn't needed.

Distributed actors can manually specify their `Transport` associated
type or pick up a default by looking for a type named
`DefaultActorTransport`. A library that vends an actor transport can
make create a public typealias `DefaultActorTransport` referring to
its transport, so importing that library and defining a distributed
actor will use that library's transport.

Introduce a type-erased `AnyActorTransport` type to provide an
explicitly dynamic actor transport. This is still an important option,
e.g., for cases where one wants to be able to dynamically change the
transport for testing or different kinds of deployment. For now, we
default to this transport in the library (via `DefaultActorTransport`),
but we may very well want to eliminate this because it will be
ambiguous with client libraries that vend their own
`DefaultActorTransport`.

ktoso commented on stdlib/public/Distributed/DistributedActor.swift:
```diff
@@ -36,6 +36,9 @@ public protocol AnyActor: Sendable, AnyObject {}
 public protocol DistributedActor:
     AnyActor, Sendable, Identifiable, Hashable, Codable {
+    /// The type of transport used to communicate with actors of this type.
+    associatedtype Transport: ActorTransport
```
👍

ktoso commented on lib/Sema/DerivedConformanceDistributedActor.cpp:
```diff
@@ -156,6 +154,36 @@ static ValueDecl *deriveDistributedActor_actorTransport(
+
+  // Return the default transport type.
+  return defaultTransportTypeDecl->getDeclaredInterfaceType();
+}
```
So... this means that we look in the current module for `DefaultActorTransport`, and that dictates the type of the `Transport` property... I remain worried about this somehow... What happens if there is no default one defined, you'd suggest that each actor must explicitly define the typealias?

I was hoping we can derive it from the init declarations:

```swift

distributed actor A {
  init(t: MyTransport) {}
  // therefore:
  typealias Transport = MyTransport
}
```

and also once all transport types must "match":

```swift
distributed actor A {
  init(t: MyTransport) {}
  init(number: Int, other: OtherTransport) {}
  // ERROR: distributed actor initializers must accept 
  // the same transport type in all initializers
```

The `DefaultActorTransport` we could still keep though! It's really be the default, but this init declaration dance I wonder if would not be worth it...? Maybe let's ignore for now hm...

DougGregor replied:
As I've noted in my newly-written comment, I expect that any library providing a transport would provide a `DefaultActorTransport` typealias pointing at that transport, so importing the library + defining a distributed actor would use that transport. Inference from `init(transport:...)` would also be possible.

ktoso commented on stdlib/public/Distributed/DistributedActor.swift:
```diff
@@ -100,9 +103,9 @@ extension CodingUserInfoKey {
 extension DistributedActor {
   nonisolated public init(from decoder: Decoder) throws {
-    guard let transport = decoder.userInfo[.actorTransportKey] as? ActorTransport else {
+    guard let transport = decoder.userInfo[.actorTransportKey] as? Transport else {
```
Hmmm... I see, this is more restrictive but I think it'll be fine -- realistically people should not be mixing completely random transports and sending an actor from one transport to another transport, I suppose...

DougGregor replied:
Oh, that's a good point, this might be a bug! We probably need `AnyActorTransport` to be `Codable` over top of the underlying transport, so we can decide it.

ktoso replied:
Nah, transports are not intended to be Codable; This change seems fine.

I was thinking about a weird use-case that to be honest we don't have to support IMHO -- "mixing transports" which is pretty complicated and I don't think we really need it in any real use-case to be honest.

Here is the imagined problem:
- you could receive a `websocket://12.3.4:####/greeter` identified actor... the `WebSocketTransport` decodes this fine...
- you are ALSO part of the cluster, and send the same ref to some cluster actor...
- encoding just works, it's Codable by the websocket address...
- the recipient system, attempts to decode it
  - puts self into `.actorTransportKey`, that's the Cluster transport
  - that as? will fail, since this is `Cluster` and not `WebSocket` transport after all

To be honest this is likely  _good_, you can't just randomly share refs like that as the connections and forwarding take a role in all that too... I think this is okey to keep as such runtime `"failed to decode websocket:/// actor using Cluster transport" it makes total sense and is easily fixed by developers with a "forwarder" actor if they need such chain of messages.


Why all of this doesn't really matter:
- the connection of course didn't "move" with the actor as we sent it over like that...
- so it does not make sense for that other system to resolve the "actual connection backed actor", since it can't, the connection is on the first node -- that accepted that websocket actor.

So yeah, such weird thing could happen, but honestly I think it's a weird edge case, and by means of our transports being explicit in the types -- we'd see that "huh, why am I sending this websocket actor to the cluster, does this make sense?" One can spin up a simple "forwarder" actor if we needed to support that.

ktoso reviewed (commented):
Overall looking good, definitely nice to have the typealias explicitly like that; while still being able to support existential if we need to hmmm

DougGregor committed:
Make the actor transport an associated type of the DistributedActor protocol.

Eliminate the required use of existentials in distributed actors by
introducing the `Transport` associated type into the
`DistributedActor` protocol. Each distributed actor has a known
(concrete) actor transport type, reducing storage requirements and
transport dynamism when it isn't needed.

Distributed actors can manually specify their `Transport` associated
type or pick up a default by looking for a type named
`DefaultActorTransport`. A library that vends an actor transport can
make create a public typealias `DefaultActorTransport` referring to
its transport, so importing that library and defining a distributed
actor will use that library's transport.

Introduce a type-erased `AnyActorTransport` type to provide an
explicitly dynamic actor transport. This is still an important option,
e.g., for cases where one wants to be able to dynamically change the
transport for testing or different kinds of deployment. For now, we
default to this transport in the library (via `DefaultActorTransport`),
but we may very well want to eliminate this because it will be
ambiguous with client libraries that vend their own
`DefaultActorTransport`.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please build toolchain

ktoso commented:
Could be real issues?

DougGregor committed:
Fix a few tests for the actor transport associated type

DougGregor commented:
@swift-ci please test

ktoso reviewed (approved):
Fantastic, LGTM!

DougGregor commented:
I have a generalization of this PR that also de-existential-ifies the actor identity. I assume we want that on top of this, but didn't want to mess with this directly, so it's over at https://github.com/apple/swift/pull/40033.

xedin commented on lib/SILGen/SILGenDistributed.cpp:
```diff
@@ -45,22 +45,45 @@ static VarDecl* lookupProperty(NominalTypeDecl *ty, DeclName name) {
+    VarDecl *property) {
+  Type formalType = SGF.F.mapTypeIntoContext(property->getInterfaceType());
+  SILType loweredType = SGF.getLoweredType(formalType).getAddressType();
+#if false
```
Is this just a left-over from debugging?

ktoso replied:
Whoop, I'll remove that in a follow up

ktoso reviewed (approved):
Merging and I'll remove the #if false and the Default typealias in a followup now.
